File: python/thunder_rpc/value.py
"""The 8-variant value model and the ``Request``/``Response`` frame bodies.

Mirrors ``thunder-wire``'s ``value.rs`` (WIRE-001/002): externally-tagged
encoding is handled by :mod:`thunder_rpc.wire`; this module is the pure data
model. ``Map`` is an insertion-ordered pair list because keys may be any
value (WIRE-002).
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Iterable

#: Negotiated wire protocol version. v1 is the only version anywhere (WIRE-004).
WIRE_VERSION = 1

#: Reserved frame id for server-initiated push frames (WIRE-005).
PUSH_ID = 0xFFFF_FFFF

#: Default frame-body cap: 64 MiB, checked before allocation (WIRE-020).
DEFAULT_MAX_FRAME_BYTES = 64 * 1024 * 1024

_I64_MIN = -(2**63)
_I64_MAX = 2**63 - 1
_U32_MAX = 2**32 - 1

#: The 8 wire value kinds (WIRE-002).
KINDS = ("null", "bool", "int", "float", "bytes", "str", "array", "map")


@dataclass(frozen=True)
class Value:
    """One wire value — byte-compatible with ``SynapValue`` / ``NexusValue`` /
    ``VectorizerValue`` (WIRE-002).

    Construct through the factories (``Value.int(42)``, ``Value.map(pairs)``,
    ...) which validate and normalize; ``kind`` is one of :data:`KINDS`.
    """

    kind: str
    value: Any

    # -- factories ----------------------------------------------------------

    @staticmethod
    def null() -> Value:
        """SQL NULL / nil."""
        return Value("null", None)

    @staticmethod
    def bool(value: bool) -> Value:  # noqa: A003 - mirrors the wire tag names
        if type(value) is not bool:
            raise TypeError(f"Value.bool needs a bool, got {type(value).__name__}")
        return Value("bool", value)

    @staticmethod
    def int(value: int) -> Value:
        if type(value) is not int:
            raise TypeError(f"Value.int needs an int, got {type(value).__name__}")
        if not _I64_MIN <= value <= _I64_MAX:
            raise ValueError(f"Value.int must fit i64, got {value}")
        return Value("int", value)

    @staticmethod
    def float(value: float) -> Value:
        if type(value) is bool or not isinstance(value, (int, float)):
            raise TypeError(f"Value.float needs a float, got {type(value).__name__}")
        return Value("float", float(value))

    @staticmethod
    def bytes(data: bytes) -> Value:
        if not isinstance(data, (bytes, bytearray, memoryview)):
            raise TypeError(f"Value.bytes needs bytes, got {type(data).__name__}")
        return Value("bytes", bytes(data))

    @staticmethod
    def str(text: str) -> Value:
        if not isinstance(text, str):
            raise TypeError(f"Value.str needs a str, got {type(text).__name__}")
        return Value("str", text)

    @staticmethod
    def array(items: Iterable[Value]) -> Value:
        items = tuple(items)
        for item in items:
            if not isinstance(item, Value):
                raise TypeError(
                    f"Value.array items must be Value, got {type(item).__name__}"
                )
        return Value("array", items)

    @staticmethod
    def map(pairs: Iterable[tuple[Value, Value]]) -> Value:
        normalized = []
        for pair in pairs:
            key, val = pair
            if not isinstance(key, Value) or not isinstance(val, Value):
                raise TypeError("Value.map pairs must be (Value, Value) tuples")
            normalized.append((key, val))
        return Value("map", tuple(normalized))

    # -- accessors (mirror thunder-wire's Value accessors) -------------------

    def as_str(self) -> str | None:
        """Extract the inner string."""
        return self.value if self.kind == "str" else None

    def as_bytes(self) -> bytes | None:
        """Extract bytes (also accepts ``Str`` as UTF-8 bytes)."""
        if self.kind == "bytes":
            return self.value
        if self.kind == "str":
            return self.value.encode("utf-8")
        return None

    def as_int(self) -> int | None:
        """Extract an integer."""
        return self.value if self.kind == "int" else None

    def as_float(self) -> float | None:
        """Extract a float (accepts ``Int`` widened to float)."""
        if self.kind == "float":
            return self.value
        if self.kind == "int":
            return float(self.value)
        return None

    def as_bool(self) -> bool | None:
        """Extract a bool."""
        return self.value if self.kind == "bool" else None

    def as_array(self) -> tuple[Value, ...] | None:
        """Extract the array items."""
        return self.value if self.kind == "array" else None

    def as_map(self) -> tuple[tuple[Value, Value], ...] | None:
        """Extract the map pairs."""
        return self.value if self.kind == "map" else None

    def map_get(self, key: str) -> Value | None:
        """Look up a string key in a ``Map`` value."""
        pairs = self.as_map()
        if pairs is None:
            return None
        for k, v in pairs:
            if k.as_str() == key:
                return v
        return None

    def is_null(self) -> bool:
        """True for ``Value.null()``."""
        return self.kind == "null"


def _check_id(frame_id: int) -> None:
    if type(frame_id) is not int or not 0 <= frame_id <= _U32_MAX:
        raise ValueError(f"frame id must be a u32, got {frame_id!r}")


@dataclass(frozen=True)
class Request:
    """One RPC request (WIRE-001). ``id`` is client-chosen and echoed back;
    many requests multiplex over one connection. Serialized as an array
    (WIRE-012); map-shaped requests decode too (WIRE-013).
    """

    id: int
    command: str
    args: tuple[Value, ...] = ()

    def __post_init__(self) -> None:
        _check_id(self.id)
        if not isinstance(self.command, str):
            raise TypeError(f"command must be a str, got {type(self.command).__name__}")
        args = tuple(self.args)
        for arg in args:
            if not isinstance(arg, Value):
                raise TypeError(f"args items must be Value, got {type(arg).__name__}")
        object.__setattr__(self, "args", args)


@dataclass(frozen=True)
class Response:
    """One RPC response (WIRE-001). Exactly one of ``ok`` / ``err`` is set;
    v1 carries no structured error object — conventions are prefix-based and
    profile-driven (WIRE-040).
    """

    id: int
    ok: Value | None = None
    err: str | None = None

    def __post_init__(self) -> None:
        _check_id(self.id)
        if (self.ok is None) == (self.err is None):
            raise ValueError("Response needs exactly one of ok/err")
        if self.ok is not None and not isinstance(self.ok, Value):
            raise TypeError(f"ok must be a Value, got {type(self.ok).__name__}")
        if self.err is not None and not isinstance(self.err, str):
            raise TypeError(f"err must be a str, got {type(self.err).__name__}")

File: rust/thunder-wire/src/value.rs
//! The 8-variant value model and the `Request`/`Response` frames.
//!
//! Externally-tagged encoding (rmp-serde default): unit variants serialize
//! as a bare string (`"Null"`), payload variants as a single-key map
//! (`{"Int": 42}`). `Response.result` is a serde `Result`, so a successful
//! string reply nests two one-key maps: `{"Ok": {"Str": "PONG"}}` — pinned
//! by the conformance corpus.

use serde::{Deserialize, Serialize};

/// The wire value model — byte-compatible with `SynapValue` /
/// `NexusValue` / `VectorizerValue` (WIRE-002).
///
/// `Map` is an insertion-ordered pair list because keys may be any value.
#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub enum Value {
    /// SQL NULL / nil.
    Null,
    Bool(bool),
    Int(i64),
    Float(f64),
    /// Raw bytes. Emitted as MessagePack **bin** (WIRE-010); the legacy
    /// int-array form decodes too (WIRE-011) via `serde_bytes`' visitor.
    Bytes(#[serde(with = "serde_bytes")] Vec<u8>),
    Str(String),
    Array(Vec<Value>),
    Map(Vec<(Value, Value)>),
}

impl Value {
    /// Extract the inner string slice.
    pub fn as_str(&self) -> Option<&str> {
        match self {
            Self::Str(s) => Some(s.as_str()),
            _ => None,
        }
    }

    /// Extract bytes (also accepts `Str` as UTF-8 bytes).
    pub fn as_bytes(&self) -> Option<&[u8]> {
        match self {
            Self::Bytes(b) => Some(b.as_slice()),
            Self::Str(s) => Some(s.as_bytes()),
            _ => None,
        }
    }

    /// Extract an integer.
    pub fn as_int(&self) -> Option<i64> {
        match self {
            Self::Int(i) => Some(*i),
            _ => None,
        }
    }

    /// Extract a float (accepts `Int` widened to `f64`).
    pub fn as_float(&self) -> Option<f64> {
        match self {
            Self::Float(f) => Some(*f),
            Self::Int(i) => Some(*i as f64),
            _ => None,
        }
    }

    /// Extract a bool.
    pub fn as_bool(&self) -> Option<bool> {
        match self {
            Self::Bool(b) => Some(*b),
            _ => None,
        }
    }

    /// Extract the array items.
    pub fn as_array(&self) -> Option<&[Value]> {
        match self {
            Self::Array(items) => Some(items.as_slice()),
            _ => None,
        }
    }

    /// Extract the map pairs.
    pub fn as_map(&self) -> Option<&[(Value, Value)]> {
        match self {
            Self::Map(pairs) => Some(pairs.as_slice()),
            _ => None,
        }
    }

    /// Look up a string key in a `Map` value.
    pub fn map_get(&self, key: &str) -> Option<&Value> {
        self.as_map()?
            .iter()
            .find(|(k, _)| k.as_str() == Some(key))
            .map(|(_, v)| v)
    }

    /// True for `Value::Null`.
    pub fn is_null(&self) -> bool {
        matches!(self, Self::Null)
    }
}

impl From<bool> for Value {
    fn from(b: bool) -> Self {
        Self::Bool(b)
    }
}
impl From<i64> for Value {
    fn from(i: i64) -> Self {
        Self::Int(i)
    }
}
impl From<f64> for Value {
    fn from(f: f64) -> Self {
        Self::Float(f)
    }
}
impl From<String> for Value {
    fn from(s: String) -> Self {
        Self::Str(s)
    }
}
impl From<&str> for Value {
    fn from(s: &str) -> Self {
        Self::Str(s.to_owned())
    }
}
impl From<Vec<u8>> for Value {
    fn from(b: Vec<u8>) -> Self {
        Self::Bytes(b)
    }
}
impl From<Vec<Value>> for Value {
    fn from(items: Vec<Value>) -> Self {
        Self::Array(items)
    }
}

/// One RPC request (WIRE-001). `id` is client-chosen and echoed back;
/// many requests multiplex over one connection. Serialized as an array
/// (WIRE-012); map-shaped requests decode too (WIRE-013).
#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct Request {
    pub id: u32,
    pub command: String,
    pub args: Vec<Value>,
}

/// One RPC response (WIRE-001). `result` is `Ok(value)` or an error
/// string; v1 carries no structured error object — conventions are
/// prefix-based and profile-driven (WIRE-040).
#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct Response {
    pub id: u32,
    pub result: Result<Value, String>,
}

impl Response {
    /// Success response.
    pub fn ok(id: u32, value: Value) -> Self {
        Self {
            id,
            result: Ok(value),
        }
    }

    /// Error response with the verbatim error string.
    pub fn err(id: u32, message: impl Into<String>) -> Self {
        Self {
            id,
            result: Err(message.into()),
        }
    }
}
